tools/ddrgen/src/scanners/dwarf/AixDwarfParser.cpp

```diff
@@ -441,6 +720,618 @@ parseDwarfInfo(const char *line)
+									if (ID > 0) {
+										ID = ID + _fileID;
+									}
+									nestedClassesToPopulate.push_back(make_pair<double, Dwarf_Die> ((double)ID, (Dwarf_Die)currentDie));
+								}
+
+								if (DW_DLV_OK == ret) {
+									/* Set the member as a child of the parent die */
+									if (NULL == lastChild) {
+										currentDie->_child = newDie;
+									} else {
+										lastChild->_sibling = newDie;
+										newDie->_previous = lastChild;
+									}
+									lastChild = newDie;
+								} else {
+									/* To prevent a memory leak, delete the new die */
+									deleteDie(newDie);
+								}
+							}
+						}
+					}
+				}
+			}
+		} else {
+			ret = DW_DLV_ERROR;
+			setError(error, DW_DLE_IA);
+		}
+	} else {
+		ret = DW_DLV_ERROR;
+		setError(error, DW_DLE_IA);
+	}
+	return ret;
+}
+/**
+ * Parses data passed in as a typedef.
+ * param[in] data: The data to parse, should be in the format t(TYPE_ID)=(TYPE)
+ * param[in] dieName: The name of the typedef
+ * param[out] currentDie: the DIE to populate with the parsed data.
+ */
+static int
+parseTypeDef(const string data,
+	const string dieName,
+	Dwarf_Die currentDie,
+	Dwarf_Error *error)
+{
+	int ret = DW_DLV_OK;
+
+	/* Set attributes for name, type, declaring file */
+	Dwarf_Attribute type = new Dwarf_Attribute_s();
+	Dwarf_Attribute name  = new Dwarf_Attribute_s();
+	Dwarf_Attribute declFile = new Dwarf_Attribute_s();
+
+	if ((NULL == type) || (NULL == name) || (NULL == declFile)) {
+		ret = DW_DLV_ERROR;
+		setError(error, DW_DLE_MAF);
+	} else {
+		type->_type = DW_AT_type;
+		type->_form = DW_FORM_ref1;
+		type->_nextAttr = name;
+		/* If the type's typeID is built-in, leave it as is, otherwise, add the current file's ID to make it unique */
+		double ID = toInt(data);
+		if (ID < 0) {
+			type->_sdata = ID;
+		} else {
+			type->_sdata = ID + _fileID;
+		}
+		type->_udata = 0;
+		type->_refdata = 0;
+		type->_ref = NULL;
+
+		/* Add type to the list of refs to populate on a second pass */
+		refsToPopulate.push_back(type);
+
+		name->_type = DW_AT_name;
+		name->_form = DW_FORM_string;
+		name->_nextAttr = declFile;
+		name->_sdata = 0;
+		name->_udata = 0;
+		name->_stringdata = strdup(dieName.c_str());
+		name->_refdata = 0;
+		name->_ref = NULL;
+
+		declFile->_type = DW_AT_decl_file;
+		declFile->_form = DW_FORM_string;
+		declFile->_nextAttr = NULL;
+		declFile->_sdata = 0;
+		declFile->_udata = 0;
+		if (!Dwarf_CU_Context::_fileList.empty()) {
+			declFile->_stringdata = strdup(Dwarf_CU_Context::_fileList.back().c_str());
+		}
+		declFile->_refdata = 0;
+		declFile->_ref = NULL;
+
+		currentDie->_attribute = type;
+	}
+	return ret;
+}
+
+/**
+ * Looks up and populates the reference attributes which were added while populating Dwarf DIEs
+ */
+static void
+populateAttributeReferences()
+{
+	/* Iterate through the vector containing all attributes to be populated*/
+	for (unsigned int i = 0; i < refsToPopulate.size(); i++) {
```

Comment:
refsToPopulate would make more sense as vector of pairs or some such collection. That would allow you to store a pair of values: an attribute which needs to be set, and an ID for what it needs to be set to. The _sdata of the attribute should not be used for this purpose. From the perspective of the DwarfScanner or someone consuming the Dwarf API, its not obvious why an attribute has 2 values or what the garbage value could mean.

As is, the code assumes that the attribute structure is defined such that both the _sdata and _ref are available, separate locations in memory which can be used simultaneously. While this is true for the current implementation, th Dwarf spec for attributes is such that only one should ever be set.